### PR TITLE
Refactor WriteImpl (pipeline write part 1)

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Refactor WriteImpl() so when I plug-in the pipeline write code (which is
an alternative approach for WriteThread), some of the logic can be
reuse. I split out the following methods from WriteImpl():

* PreprocessWrite()
* HandleWALFull() (previous MaybeFlushColumnFamilies())
* HandleWriteBufferFull()
* WriteToWAL()

Also adding a constructor to WriteThread::Writer, and move WriteContext into db_impl.h.
No real logic change in this patch.

Test Plan:
Make sure existing tests pass.

### CODE REVIEW

File: db/db_impl.cc [hunk 1]
```diff
@@ -4687,10 +4669,6 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
     // we are a non-leader in a parallel group
     PERF_TIMER_GUARD(write_memtable_time);
 
-    if (log_used != nullptr) {
-      *log_used = w.log_used;
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: This is done on STATE_COMPLETED (if it is a follwer) and end of WriteImpl (if it is a leader). Removing duplicate.

File: db/db_impl.cc [hunk 2]
```diff
@@ -106,20 +106,6 @@ const std::string kDefaultColumnFamilyName("default");
 
 void DumpRocksDBBuildVersion(Logger * log);
 
-struct DBImpl::WriteContext {
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: Moved to db_impl.h

File: db/db_impl.cc [hunk 3]
```diff
@@ -2572,15 +2569,6 @@ Status DBImpl::SetDBOptions(
 #endif  // ROCKSDB_LITE
 }
 
-Status DBImpl::PersistOptions() {
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: removed and inline the content to all call sites.

File: db/db_impl.cc [hunk 4]
```diff
@@ -4717,123 +4695,35 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
-    RecordTick(stats_, WRITE_DONE_BY_OTHER);
     return w.FinalStatus();
   }
   // else we are the leader of the write batch group
   assert(w.state == WriteThread::STATE_GROUP_LEADER);
 
-  WriteContext context;
-  mutex_.Lock();
-
-  if (!write_options.disableWAL) {
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: moved to later in this WriteImpl().

File: db/db_impl.cc [hunk 5]
```diff
@@ -4717,123 +4695,35 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
   // This is how a write job could be done by the other writer.
-  assert(!single_column_family_mode_ ||
-         versions_->GetColumnFamilySet()->NumberOfColumnFamilies() == 1);
-
-  if (UNLIKELY(!single_column_family_mode_ &&
-               total_log_size_ > GetMaxTotalWalSize())) {
-    MaybeFlushColumnFamilies();
-  }
-  if (UNLIKELY(write_buffer_manager_->ShouldFlush())) {
-    // Before a new memtable is added in SwitchMemtable(),
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: moved to HandleWriteBufferFull()

File: db/db_impl.cc [hunk 6]
```diff
@@ -4717,123 +4695,35 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
-      status = SwitchMemtable(cfd_picked, &context);
-      if (status.ok()) {
-        cfd_picked->imm()->FlushRequested();
-        SchedulePendingFlush(cfd_picked);
-        MaybeScheduleFlushOrCompaction();
-      }
-    }
-  }
-
-  if (UNLIKELY(status.ok() && !bg_error_.ok())) {
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: The following logic is moved to PreprocessWrite().

File: db/db_impl.cc [hunk 7]
```diff
@@ -4874,110 +4764,44 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
+      RecordTick(stats_, WRITE_DONE_BY_OTHER, write_done_by_other);
+    }
     MeasureTime(stats_, BYTES_PER_WRITE, total_byte_size);
-    PERF_TIMER_STOP(write_pre_and_post_process_time);
 
     if (write_options.disableWAL) {
       has_unpersisted_data_.store(true, std::memory_order_relaxed);
     }
 
-    uint64_t log_size = 0;
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: The following logic is moved to WriteToWAL().

File: db/db_impl.cc [hunk 8]
```diff
@@ -5096,8 +5052,6 @@ void DBImpl::MaybeFlushColumnFamilies() {
     alive_log_files_.begin()->getting_flushed = true;
   }
 
-  WriteContext context;
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: WriteContext is passing from WriteImpl() instead and destruct without holding mutex lock.

File: db/db_impl.cc (hunk 3 above)
Comment (maysamyabandeh) [inline db/db_impl.cc]: out of curiosity, why?

File: db/db_impl.cc (hunk 8 above)
Comment (maysamyabandeh) [inline db/db_impl.cc]: How is that we can replace these two WriteContexts with no sideeffect?

File: db/db_impl.cc [hunk 9]
```diff
@@ -5063,13 +4887,145 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
...
 
-void DBImpl::MaybeFlushColumnFamilies() {
+Status DBImpl::PreprocessWrite(const WriteOptions& write_options,
+                               bool need_log_sync, bool* logs_getting_synced,
+                               WriteContext* write_context) {
   mutex_.AssertHeld();
+  assert(write_context != nullptr && logs_getting_synced != nullptr);
+  Status status;
 
-  if (alive_log_files_.begin()->getting_flushed) {
```
Comment (maysamyabandeh) [inline db/db_impl.cc]: Why is that this check is removed?

File: db/db_impl.cc [hunk 10]
```diff
@@ -4717,123 +4695,35 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       *log_used = w.log_used;
     }
     // write is complete and leader has updated sequence
-    RecordTick(stats_, WRITE_DONE_BY_OTHER);
```
Comment (maysamyabandeh) [inline db/db_impl.cc]: Why is this removed?

File: db/db_impl.cc (hunk 4 above)
Comment (maysamyabandeh) [inline db/db_impl.cc]: Where? I do not see it in WriteImpl!

File: db/db_impl.cc (hunk 3 above)
Comment (yiwu-arbug) [inline db/db_impl.cc]: In DBSetOptions I would like to call HandleWALFull() between EnterUnbatch/ExitUnbatch, making this method useless.

File: db/db_impl.cc (hunk 10 above)
Comment (yiwu-arbug) [inline db/db_impl.cc]: The logic is moved to WriteToWAL().

File: db/db_impl.cc (hunk 4 above)
Comment (yiwu-arbug) [inline db/db_impl.cc]: Sorry. It is in WriteToWAL().

File: db/db_impl.cc (hunk 9 above)
Comment (yiwu-arbug) [inline db/db_impl.cc]: Good catch! Will fix.

File: db/db_impl.cc (hunk 8 above)
Comment (yiwu-arbug) [inline db/db_impl.cc]: The purpose of WriteContext is to store pointers to superversions_to_free and memtables_to_free in order to deallocate them after release mutex lock. Seems during previous code changes the WriteContext is deallocate with mutex locked, and I'm fixing it.
